### DISCUSSION TITLE
Specifying output types to improve TypeScript performance?

### DISCUSSION THREAD

**Discussion by serg06** [trpc/trpc#3574] [Q&A] [answered]
Hello! I'm using tRPC 10 with TypeScript 4.9.4. I have a monorepo with a client and a server.

My editor's auto-complete is pretty slow; I'd like to speed it up by explicitly specifying trpc's output types. I know I can use `.output()` + `zod` to *validate* output types at run-time, but I don't need that; I only need to declare the output type.

Is there any way to do that?

**Comment by KATT** [comment, marked as answer]
You should be able to do it on the callback like on any other callback function. `: Promise<Thing>`

This issue might be also be solved by splitting up your app more

**Reply by serg06** [reply]
Ohh wow that works, thank you.

**Reply by KATT** [reply]
If you split up your backend tRPC router in several packages it should give you a lot of perf gains too, you can do this either by using a monorepo like nx/turbo or by using project references

References:

- https://github.com/microsoft/TypeScript/wiki/Performance
- https://www.typescriptlang.org/docs/handbook/project-references.html

**Reply by serg06** [reply]
Thanks for the suggestion. We use turbo; I'll start by moving the router to its own package.

I read through project references but don't understand how they could help. They only seem to speed up `tsc` performance, not editor performance.

**Reply by chrisvander** [reply]
Editor uses `tsc` in incremental mode under the hood to parse types. Any improvements to TypeScript compilation will be reflected in editor autocomplete speed.

**Reply by dBianchii** [reply]
> If you split up your backend tRPC router in several packages it should give you a lot of perf gains too, you can do this either by using a monorepo like nx/turbo or by using project references
> 
> References:
> 
> * https://github.com/microsoft/TypeScript/wiki/Performance
> * https://www.typescriptlang.org/docs/handbook/project-references.html

Hi @KATT , I am in a situation where I am losing a LOT of performance by using trpc. So much so that I am actually even unfortunately considering moving to another backend solution :crying_cat_face: 
I love tRPC and want to keep using. But it's becoming very hard to justify because of my LSP performance. Do you have a link to an example of how to _split up my trpc_ as you said?

**Comment by jonsoku-dev** [comment]
It's so slow that you can't use it..
With DX like this, it's so painful because it's slow enough to give up type checking, which is the advantage of trpc.
